### PR TITLE
Use generic for file loading in FileMiddleware

### PR DESCRIPTION
**Pull request by adam-fowler**
This will allow for writing custom file loaders that cache file contents or use another source than the local file system for files

### CODE REVIEW

File: Sources/Hummingbird/Files/LocalFileSystem.swift [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        do {
+            let lstat = try await self.fileIO.fileIO.lstat(path: path)
+            let isFolder = (lstat.st_mode & S_IFMT) == S_IFDIR
+            let modificationDate = Double(lstat.st_mtimespec.tv_sec) + (Double(lstat.st_mtimespec.tv_nsec) / 1_000_000_000.0)
+            return .init(
+                isFolder: isFolder,
+                size: numericCast(lstat.st_size),
+                modificationDate: Date(timeIntervalSince1970: modificationDate)
+            )
+        } catch {
```
Comment (Joannis) [inline Sources/Hummingbird/Files/LocalFileSystem.swift:75]: Is there a known error that would be thrown if the file does not exist?

**Review by Joannis — COMMENTED** [review]
I like the symbol names a lot, but want to check if the `FileProvider` protocol has all the functions we would need

File: Sources/Hummingbird/Files/LocalFileSystem.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/Hummingbird/Files/LocalFileSystem.swift:75]: It's a `NIOCore.IOError(2)` but this could change when we move to NIOFileSystem.

**Comment by adam-fowler** [comment]
> I like the symbol names a lot, but want to check if the `FileProvider` protocol has all the functions we would need

Yeah that is a concern. The FileMiddleware covers pretty much all eventualities. The main possibility would be we need another file attribute.

File: Sources/Hummingbird/Files/LocalFileSystem.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    public struct FileAttributes: Sendable, FileMiddlewareFileAttributes {
+        /// Is file a folder
+        public let isFolder: Bool
+        /// Size of file
+        public let size: Int
+        /// Last time file was modified
+        public let modificationDate: Date
+
+        /// Initialize FileAttributes
+        public init(isFolder: Bool, size: Int, modificationDate: Date) {
```
Comment (Joannis) [inline Sources/Hummingbird/Files/LocalFileSystem.swift:31]: Can this be made `internal` now?

**Review by Joannis — CHANGES_REQUESTED** [review]
Only one change, then it's fine

File: Sources/Hummingbird/Files/LocalFileSystem.swift (hunk 2 above)
Comment (Joannis) [reply Sources/Hummingbird/Files/LocalFileSystem.swift:31]: I don't think we want to lock ourselves in here

Comment (adam-fowler) [reply Sources/Hummingbird/Files/LocalFileSystem.swift:31]: yeah that should never have been public

Comment (adam-fowler) [reply Sources/Hummingbird/Files/LocalFileSystem.swift]: done

**Review by Joannis — APPROVED** [review]

**Pull request merged**